UCCSD with density fitting

hczhai opened the pull request:
This pull request implements the memory efficient version of DF-UCCSD, by avoiding the explicit construction of ``vvvv`` (similar to ``cc.dfccsd.RCCSD``).

When ``uhf`` has ``with_df``, ``cc.CCSD(uhf)`` previously dispatched to ``cc.uccsd.UCCSD(uhf)`` (with explicit ``vvvv``, see https://github.com/pyscf/pyscf/pull/1158). After this PR, ``cc.CCSD(uhf)`` will dispatch to ``cc.dfuccsd.UCCSD(uhf)`` (without explicit ``vvvv``), which matches the behavior of DF-RCCSD. If one explicitly call ``cc.uccsd.UCCSD(uhf)``, ``vvvv`` will still be explicitly constructed, and a warning will be printed if ``uhf`` has ``with_df``.

Benchmark using C<sub>6</sub>H<sub>6</sub> ([geometry](https://doi.org/10.1021/acs.jpclett.0c02621)) with basis ``cc-pvqz/cc-pvqz-ri`` and ``mol.max_memory = 117GB``. On 64-core icelake. Number of orbitals ``nocc = (21, 21), nmo = (510, 510), naux = 1122``. Wall time and memory usage:

| Implementation  | AO to MO time | UCCSD per cycle time | Peak memory | E(UCCSD) |
| :--- | -------------: | ----: | ----: | ----: |
| ``cc.uccsd.UCCSD(uhf)``  | 141 sec | 710 sec | 114 GB | -231.9921598100943 Ha |
| ``cc.dfuccsd.UCCSD(uhf)``  | 49 sec  | 760 sec | 56 GB | -231.9921598100928 Ha |

jeanwsr commented:
There's a few changes in uccsd.py replacing np.einsum with np.einsum(optimize=True). 

- Should we apply this to ccsd.py as well?
- Is lib.einsum better in these cases? (maybe related to #1699)

I believe some of the `np.einsum() without optimize` here are used because numpy's `optimize` will not take effect for them, but some of them may be actually worth an `optimize=True`.

hczhai commented:
Hi @jeanwsr,

Thanks for the insightful comments.

> Is ``lib.einsum`` better in these cases?

Currently there are 45 ``numpy.einsum`` calls in ``uccsd.py``, and there are 17 unique contraction patterns among these ``numpy.einsum`` calls. So I wrote a script to help decide which version of ``einsum`` should one use for each of these contraction patterns.

<details>

<summary>Testing script and output</summary>

```python3
import numpy as np, time, itertools
from pyscf import lib

no, nv = 21, 489
t1 = np.random.random((no, nv))
t2 = np.random.random((no, no, nv, nv))

fvv = np.random.random((nv, nv))
foo = np.random.random((no, no))
fov = np.random.random((no, nv))

ovvv = np.random.random((no, nv, nv, nv))
ovov = np.random.random((no, nv, no, nv))
ovoo = np.random.random((no, nv, no, no))
oovv = np.random.random((no, no, nv, nv))
ovvo = np.random.random((no, nv, nv, no))

samples = [
    ('mf,mfae->ae', t1, ovvv),      # [#00] 1-4
    ('mf,mfae->ae', t1, ovoo),      # [#01] 5,7,9,10
    ('mf,mfae->ae', t1, ovov),      # [#02] 12
    ('nf,menf->me', t1, ovov),      # [#03] 6,8,11
    ('nf,niaf->ia', t1, oovv),      # [#04] 21,22
    ('nf,nfai->ia', t1, ovvo),      # [#05] 23,24
    ('ie,ae->ia', t1, fvv),         # [#06] 13,17
    ('ma,mi->ia', t1, foo),         # [#07] 14,18
    ('imea,me->ia', t2, fov),       # [#08] 15,19
    ('imae,me->ia', t2, fov),       # [#09] 16
    ('miea,me->ia', t2, fov),       # [#10] 20
    ('ia,ia', fov, t1),             # [#11] 25,26
    ('ijab,iajb', t2, ovov),        # [#12] 27,29,31,37,38,40
    ('ijab,ibja', t2, ovov),        # [#13] 28,30,39,41
    ('ia,jb,iajb', t1, t1, ovov),   # [#14] 32,34,36
    ('ia,jb,ibja', t1, t1, ovov),   # [#15] 33,35
    ('ia,jb->ijab', t1, t1),        # [#16] 42,44,45
    ('ia,jb->jiab', t1, t1),        # [#17] 43
]

f1 = lambda script, *tensors: np.einsum(script, *tensors)
f2 = lambda script, *tensors: np.einsum(script, *tensors, optimize=True)
f3 = lambda script, *tensors: lib.einsum(script, *tensors)

perms = list(itertools.permutations(range(3)))

for _ in range(2):
    for idx, (script, *tensors) in enumerate(samples):
        ts = [0, 0, 0]
        for ir in range(5):
            for i in perms[np.random.randint(len(perms))]:
                tx = time.perf_counter()
                r = [f1, f2, f3][i](script, *tensors)
                ts[i] += (ir != 0) * (time.perf_counter() - tx)
        t1, t2, t3 = ts
        use = 'EIN OPT LIB'.split()[np.argmin(ts)]
        use = 'ANY' if abs(t1 - t2) < max(0.001, 0.05 * min(*ts)) and abs(t2 - t3) < max(0.001, 0.05 * min(*ts)) else use
        use = 'OPT/LIB' if t1 > 1.05 * min(*ts) and abs(t2 - t3) < 0.05 * min(*ts) else use
        use = 'EIN/LIB' if t2 > 1.05 * min(*ts) and abs(t1 - t3) < 0.05 * min(*ts) else use
        print("CASE %2d EIN = %7.3f OPT = %7.3f LIB = %7.3f -> should use %s" % (idx, t1, t2, t3, use))
    print()
```

The output on 64-core icelake (a similar result is obtained on MacBook M3):

```
CASE  0 EIN =   9.122 OPT =  65.710 LIB =   9.124 -> should use EIN/LIB
CASE  1 EIN =   0.008 OPT =   0.059 LIB =   0.008 -> should use EIN/LIB
CASE  2 EIN =   0.277 OPT =   1.704 LIB =   0.281 -> should use EIN/LIB
CASE  3 EIN =   0.338 OPT =   0.865 LIB =   0.328 -> should use EIN/LIB
CASE  4 EIN =   0.298 OPT =   0.297 LIB =   0.297 -> should use ANY
CASE  5 EIN =   0.281 OPT =   1.704 LIB =   0.279 -> should use EIN/LIB
CASE  6 EIN =   0.016 OPT =   0.023 LIB =   0.130 -> should use EIN
CASE  7 EIN =   0.000 OPT =   0.000 LIB =   0.000 -> should use ANY
CASE  8 EIN =   0.320 OPT =   0.318 LIB =   0.318 -> should use ANY
CASE  9 EIN =   0.441 OPT =   0.441 LIB =   0.441 -> should use ANY
CASE 10 EIN =   0.351 OPT =   0.354 LIB =   0.351 -> should use ANY
CASE 11 EIN =   0.000 OPT =   0.000 LIB =   0.000 -> should use ANY
CASE 12 EIN =   0.604 OPT =   0.604 LIB =   0.604 -> should use ANY
CASE 13 EIN =   1.917 OPT =   1.916 LIB =   1.916 -> should use ANY
CASE 14 EIN =   0.780 OPT =   1.680 LIB =   0.261 -> should use LIB
CASE 15 EIN =   0.787 OPT =   0.296 LIB =   0.313 -> should use OPT
CASE 16 EIN =   1.147 OPT =   1.148 LIB =   1.149 -> should use ANY
CASE 17 EIN =   0.829 OPT =   0.830 LIB =   0.833 -> should use ANY
```

</details>

**The conclusion:** The five ``numpy.einsum`` involving three tensors should be changed to ``lib.einsum``. For the remaining 40 ``numpy.einsum``, ``optimize=True`` should not be added (no effect / make it worse). I have updated the PR accordingly.

> Should we apply this to ``ccsd.py`` as well?

Since ``ccsd.py`` does not have ``numpy.einsum`` calls with three tensors, there is nothing to be changed there.

> I believe some of the ``np.einsum() without optimize`` here are used because numpy's optimize will not take effect for them, but some of them may be actually worth an ``optimize=True``.

This is indeed what I observed. For some contractions in ``uccsd.py``, adding ``optimize=True`` can make performance significantly worse for some reason.
